### ISSUE TITLE
Uno-Choice Scriptler should accept a referenced parameter as a script parameter

### ISSUE THREAD

**Issue by imoutsatsos** [biouno/uno-choice-plugin#16]
I have noticed that using a Scriptler to provide a dynamic choice list works when the UI referenced parameter is 'hard-coded' in the script, but does not work when I try to pass it in a scipt parameter. I will document this in a bit more detail tomorrow.

Basically if we have a UI parameter named DATA_RUN it can be used as such in the Scriptler and it works. If the Scriptler has a parameter named dataRun and in the Uno-Choice Scriptlet configuration I try to assign dataRun=DATA_RUN this does not work.

This should make the script truly re-usable without dependency on specific UI parameter names.

**Comment by RoeiA** [comment]
Just wanna say, that this plugin is great!
However, I can't access a referenced parameter in my Scriptler Groovy script. My job parameter config is this:

![image](https://cloud.githubusercontent.com/assets/8946031/5626403/7e4f0514-9593-11e4-891f-3c9d8e0d50df.png)

and my script is this:

![image](https://cloud.githubusercontent.com/assets/8946031/5626434/ff70b9f8-9593-11e4-8418-3ba33443bc6a.png)

and when I build the project, the rollback_spec list stays empty, and I get this message in Jenkins log:
Jan 6, 2015 9:05:12 AM org.biouno.unochoice.AbstractScriptableParameter getChoices
WARNING: Script parameter with name 'Rollback_spec' is not an instance of java.util.Map. The parameter value is null

My jenkins version is 1.596 and uno-choice plugin version is 0.21. Am I doing anything wrong?

Thanks.

**Comment by imoutsatsos** [comment]
Hi @RoeiA 
In the current Uno-Choice version you have to reference the variables in the script with the names available in the binding. Issue #16 prevents you from assigning a binding variable to a script variable as you (and me) are assuming.

So to get this to work try in you scriptler script replacing **rollback** (the script variable) with **ROLLBACK**  (the binding variable):

```
if (ROLLBACK.equals ('No'))
```

Note that the other option is to define rollbcak in the script as follows:
```
def rollback=binding.variables.get('ROLLBACK')
```

We hope that this will change to work as expected in future improvements to the plugin!

**Comment by RoeiA** [comment]
Thanks for the quick response @imoutsatsos!

I tried both options you suggested and still it doesn't work.
with the second option I wrote in my scriptler this:

![image](https://cloud.githubusercontent.com/assets/8946031/5642702/dee40a28-9656-11e4-96d0-f35bbbf49a28.png)

(The println is just for debugging)

And the jenkins log shows this:
`Rollback = null`
`Jan 7, 2015 8:18:53 AM org.biouno.unochoice.AbstractScriptableParameter getChoices
WARNING: Script parameter with name 'Rollback_spec' is not an instance of java.util.Map. The parameter value is null`

As you can see the value of my variable is null, and I didn't change anything else.

Appreciate your help, best regards.

**Comment by imoutsatsos** [comment]
Hi @RoeiA ; I'm sorry this is still an issue.
Please, install uno-choice v 0.22 from here:
https://groups.google.com/forum/#!topic/biouno-developers/awNMG5QqYSk

and retry.
I have build a test project and it seems to work for me with v 0.22.
give it a try here: http://162.243.246.75:8080/job/issue16A

![image](https://cloud.githubusercontent.com/assets/3408226/5645381/b6d6b540-963c-11e4-8387-de2d55845ed0.png)

![image](https://cloud.githubusercontent.com/assets/3408226/5645388/c9dc7f12-963c-11e4-8be7-b6c5c231ef81.png)

![image](https://cloud.githubusercontent.com/assets/3408226/5645398/dc9d0e6e-963c-11e4-95d9-4b6051431624.png)

```
def rollbackA=binding.variables.get('ROLLBACK_A')
def rollbackB=binding.variables.get('ROLLBACK_B')
options=[]
if (rollbackA=='yes'){
  options.add('yes_choice')
}else{
  options.add('no_choice')
}

if (rollbackB=='yes'){
  options.add('yes_uno')
}else{
  options.add('no_uno')
}
return options
```

Best regards
Ioannis

**Comment by imoutsatsos** [comment]
You are welcome! Thanks for your persistence @RoeiA !

**Comment by kinow** [comment]
Fixed. Now parameters in the Scriptler configuration - in the job config - can be defined by the user (e.g. "weather" => "sunny") or can reference the name of other parameters (e.g. "brazilianState" => ${State}). 

The notation is the same used throughout Jenkins for referencing variables. One downside, is that users will have to update jobs if they were expecting that the plug-in would hard-code the plug-in values - in other words, not that we have the correct behaviour, they will have to update their jobs to reflect this change :-) a positive trouble I think. 

Thanks!

**Comment by kinow** [comment]
Deployed to our testing server @imoutsatsos . Hopefully working now.

**Comment by imoutsatsos** [comment]
Excellent. It looks good!
We should note that this really simplifies the way scriptlets get parameterized. Note that a scriptler does not have to do anything special to define variables whose name/values are assigned in the parameter configuration.

For example given the following configuration:
![untitled_clipping_011215_044109_pm](https://cloud.githubusercontent.com/assets/3408226/5711602/d0b2a244-9a79-11e4-8ada-d39c95ddfbf4.jpg)

The scriptlet does not even need to define the *brazilianState* and *weather variables*! it can just use them.
![untitled_clipping_011215_044551_pm](https://cloud.githubusercontent.com/assets/3408226/5711701/78bbda8c-9a7a-11e4-9c99-bdc5f5eca742.jpg)

Also the value of *brazilianState* can be assigned either as *$States* or *${States}*
Both seem to work.

**Comment by kinow** [comment]
Yay, nice @imoutsatsos ! You were right about fixing this. I don't use much Scriptler parameters, but now I see that it would be a big problem to maintain lots of scriptler scripts because of different parameter names. Thanks!

**Comment by kinow** [comment]
Should be fixed in the Active Choices 1.0 release. Closing.
